src/__tests__/utils.js

```diff
@@ -0,0 +1,82 @@
+/* @flow */
+
+import Client from '../index';
+
+import type {
+  RequestFactory,
+  RequestPerformer,
+  ResponseParser,
+} from '../types';
+
+type TestRequest = {
+  uri: string,
+  method?: string,
+  body?: Object,
+};
+
+type JSON = Object | Array<JSON> | number | string;
+
+type TestResponse = {
+  status: number,
+  body?: JSON,
+};
+
+export class TestRequestFactory implements RequestFactory<TestRequest> {
+  createRequest = (uri: string, method?: string, body?: Object) => ({
+    uri,
+    method,
+    body,
+  });
+}
+
+export class TestRequestPerformer
+  implements RequestPerformer<TestRequest, TestResponse> {
+  static resolve = (request: TestRequest, response: TestResponse) =>
+    new TestRequestPerformer(() => Promise.resolve(response));
+
+  static reject = (request: TestRequest, response: TestResponse) =>
+    new TestRequestPerformer(() => Promise.reject(response));
+
+  constructor(resolver: (request: TestRequest) => Promise<TestResponse>) {
+    this.resolver = resolver;
+  }
+
+  resolver: (request: TestRequest) => Promise<TestResponse>;
```

Comment:
This type signature looks similar to the type signature on line 40. Is there no way to abstract this or something?

Reply:
Good question. I think most of this was me expressing type signatures verbosely, and then filling in the method. I'll clean these up!

Reply:
So the best I could come up with here is to use a type alias (like `type PromiseResolver<Input, Output> = Input => Promise<Output>`) to avoid repeating the signature. It's a little bit cleaner.

The reason we can't really munge this down further is because we need to declare the type in both the constructor and and in the instance property declaration, and then provide the definition in actual JS code. I'll push up a change that includes the type alias, because I think it's a little cleaner.